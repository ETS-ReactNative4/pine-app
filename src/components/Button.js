import React, { Component } from 'react';
import { StyleSheet, TouchableOpacity, View, Dimensions, ActivityIndicator } from 'react-native';
import PropTypes from 'prop-types';
import { ifIphoneX } from 'react-native-iphone-x-helper';
import { LinearTextGradient } from 'react-native-text-gradient';

const windowDimensions = Dimensions.get('window');
const FULL_WIDTH = windowDimensions.width;
const DEFAULT_WIDTH = FULL_WIDTH - 80;

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#F6F6F6',
    width: DEFAULT_WIDTH,
    height: 42,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 7,
    shadowRadius: 4
  },
  fullWidth: {
    width: FULL_WIDTH,
    borderRadius: 0,
    bottom: ifIphoneX(-24, 0)
  },
  label: {
    fontFamily: 'System',
    fontWeight: '600',
    fontSize: 11,
    letterSpacing: 0.41
  },
  disabled: {
    opacity: 0.4,
    color: '#7A7A7A'
  },
  loader: {
    height: 12,
    position: 'absolute',
    top: 15,
    left: 15
  }
});

export default class Button extends Component {
  constructor(props) {
    super(...arguments);

    this.state = {
      disabled: props.disabled || false,
      loading: false,
      fullWidth: false
    };
  }

  componentDidMount() {
    /*
     * HACK: The promise from onPress might resolve and try to update state
     * after this component has been unmounted which leads to a warning.
     * One solution is to wrap the promise but this "hack" is easier in
     * this case.
     * <https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html>
     */
    this._isMounted = true;
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.disabled !== nextProps.disabled) {
      this.setState({
        disabled: nextProps.disabled
      });
    }

    if (this.props.fullWidth !== nextProps.fullWidth) {
      this.setState({
        fullWidth: nextProps.fullWidth
      });
    }
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  _onPress() {
    const onPress = this.props.onPress;
    const promise = onPress ? onPress() : null;

    if (promise instanceof Promise === false) {
      return;
    }

    this.setState({
      disabled: true,
      loading: true
    });

    promise.then(() => {
      if (!this._isMounted) {
        return;
      }

      this.setState({
        disabled: false,
        loading: false
      });
    });
  }

  render() {
    const { disabled, loading } = this.state;
    const label = loading ? this.props.loadingLabel : this.props.label;

    const buttonStyles = [
      styles.button,
      this.props.style
    ];

    let gradientTextProps = {
      colors: ['#8069DC', '#4874B3'],
      locations: [0.3, 1],
      start: { x: 0.5, y: 0 },
      end: { x: 0.5, y: 1 }
    };

    if (disabled) {
      buttonStyles.push(styles.disabled);
      gradientTextProps = {};
    }

    if (this.state.fullWidth) {
      buttonStyles.push(styles.fullWidth);
    }

    return (
      <TouchableOpacity disabled={disabled} activeOpacity={0.7} onPress={this._onPress.bind(this)}>
        <View style={buttonStyles}>
          <ActivityIndicator animating={loading} color='#7A7A7A' style={styles.loader} size='small' />
          <LinearTextGradient
            {...gradientTextProps}
            style={styles.label}
          >
            {label.toUpperCase()}
          </LinearTextGradient>
        </View>
      </TouchableOpacity>
    );
  }
}

Button.propTypes = {
  label: PropTypes.string.isRequired,
  loadingLabel: PropTypes.string,
  onPress: PropTypes.func,
  style: PropTypes.any,
  fullWidth: PropTypes.bool,
  disabled: PropTypes.bool
};
