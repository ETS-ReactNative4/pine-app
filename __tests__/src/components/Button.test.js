import React from 'react';
import renderer from 'react-test-renderer';
import Button from '../../../src/components/Button';

describe('Button', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Button label='06b5f369-ec66-4e0a-8777-90c71dec3f7b' onPress={() => {}} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly in full width', () => {
    const tree = renderer.create(
      <Button label='099349fd-1c84-415b-818f-c8c6af2ea8e7' fullWidth={true} onPress={() => {}} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly as disabled', () => {
    const tree = renderer.create(
      <Button label='15013b10-4256-4d32-b25f-a031dfb6ce95' disabled={true} onPress={() => {}} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
