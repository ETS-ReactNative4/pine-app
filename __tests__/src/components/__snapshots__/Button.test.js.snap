// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#322A51",
          "borderRadius": 7,
          "height": 46,
          "justifyContent": "center",
          "shadowRadius": 4,
          "width": 670,
        },
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={false}
      color="#D2DCFF"
      hidesWhenStopped={true}
      size="small"
      style={
        Object {
          "height": 12,
          "left": 15,
          "position": "absolute",
          "top": 15,
        }
      }
    />
    <RNLinearTextGradient
      accessible={true}
      allowFontScaling={true}
      colors={
        Array [
          4294962687,
          4292663551,
          4292736255,
        ]
      }
      ellipsizeMode="tail"
      gradientEnd={
        Object {
          "x": 1,
          "y": 1,
        }
      }
      gradientStart={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      locations={
        Array [
          0,
          0.5,
          1,
        ]
      }
      style={
        Array [
          Object {
            "color": "gray",
          },
          Object {
            "fontFamily": "System",
            "fontSize": 11,
            "fontWeight": "600",
            "letterSpacing": 0.41,
          },
        ]
      }
    >
      06B5F369-EC66-4E0A-8777-90C71DEC3F7B
    </RNLinearTextGradient>
  </View>
</View>
`;

exports[`Button renders correctly as disabled 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#322A51",
          "borderRadius": 7,
          "height": 46,
          "justifyContent": "center",
          "shadowRadius": 4,
          "width": 670,
        },
        undefined,
        Object {
          "opacity": 0.6,
        },
      ]
    }
  >
    <ActivityIndicator
      animating={false}
      color="#D2DCFF"
      hidesWhenStopped={true}
      size="small"
      style={
        Object {
          "height": 12,
          "left": 15,
          "position": "absolute",
          "top": 15,
        }
      }
    />
    <RNLinearTextGradient
      accessible={true}
      allowFontScaling={true}
      colors={
        Array [
          4294962687,
          4292663551,
          4292736255,
        ]
      }
      ellipsizeMode="tail"
      gradientEnd={
        Object {
          "x": 1,
          "y": 1,
        }
      }
      gradientStart={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      locations={
        Array [
          0,
          0.5,
          1,
        ]
      }
      style={
        Array [
          Object {
            "color": "gray",
          },
          Object {
            "fontFamily": "System",
            "fontSize": 11,
            "fontWeight": "600",
            "letterSpacing": 0.41,
          },
        ]
      }
    >
      15013B10-4256-4D32-B25F-A031DFB6CE95
    </RNLinearTextGradient>
  </View>
</View>
`;

exports[`Button renders correctly in full width 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#322A51",
          "borderRadius": 7,
          "height": 46,
          "justifyContent": "center",
          "shadowRadius": 4,
          "width": 670,
        },
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={false}
      color="#D2DCFF"
      hidesWhenStopped={true}
      size="small"
      style={
        Object {
          "height": 12,
          "left": 15,
          "position": "absolute",
          "top": 15,
        }
      }
    />
    <RNLinearTextGradient
      accessible={true}
      allowFontScaling={true}
      colors={
        Array [
          4294962687,
          4292663551,
          4292736255,
        ]
      }
      ellipsizeMode="tail"
      gradientEnd={
        Object {
          "x": 1,
          "y": 1,
        }
      }
      gradientStart={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      locations={
        Array [
          0,
          0.5,
          1,
        ]
      }
      style={
        Array [
          Object {
            "color": "gray",
          },
          Object {
            "fontFamily": "System",
            "fontSize": 11,
            "fontWeight": "600",
            "letterSpacing": 0.41,
          },
        ]
      }
    >
      099349FD-1C84-415B-818F-C8C6AF2EA8E7
    </RNLinearTextGradient>
  </View>
</View>
`;
